### DISCUSSION TITLE
logout project Spring doesn't work

### DISCUSSION THREAD

**Discussion by chris-charisma** [keycloak/keycloak#28584] [Q&A]
hello, I have an application that allows you to connect via github, google and keycloak. Logout does not work, however the JSESSIONID is deleted at logout. What do you think is the problem?

Security class:

` @Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)

public class SecurityConfig  {
 
    private final ClientRegistrationRepository clientRegistrationRepository;
  
    public  SecurityConfig(ClientRegistrationRepository clientRegistrationRepository){
        this.clientRegistrationRepository=clientRegistrationRepository;
    }

    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http.csrf(Customizer.withDefaults())//config par def.
                .authorizeHttpRequests(ar-> ar.anyRequest().authenticated())//toutes requetes necessite une authentification
                .authorizeHttpRequests(ar->ar.requestMatchers("/","/webjars/**","/h2-console/**",  "/oauth2Login/**").permitAll())
                .oauth2Login(al->al.loginPage("/oauth2Login"))
                .logout((logout)->logout
                        .logoutSuccessHandler(oidcLogoutSuccessHandler())
                        .logoutSuccessUrl("/").permitAll()
                        .clearAuthentication(true)
                        .deleteCookies("JSESSIONID","jti", "AUTH_SESSION_ID", "KEYCLOAK_SESSION", "KEYCLOAK_IDENTITY", "KC_RESTART"))
                        .exceptionHandling(eh->eh.accessDeniedPage("/notAuthorized"));
        return http.build();

    }

    private OidcClientInitiatedLogoutSuccessHandler oidcLogoutSuccessHandler() {
        final OidcClientInitiatedLogoutSuccessHandler oidcClientInitiatedLogoutSuccessHandler=
                new OidcClientInitiatedLogoutSuccessHandler(this.clientRegistrationRepository);
        oidcClientInitiatedLogoutSuccessHandler.setPostLogoutRedirectUri("{baseUrl}?logoutsuccess=true");
        return oidcClientInitiatedLogoutSuccessHandler;
    }
    @Bean
    public GrantedAuthoritiesMapper userAuthoritiesMapper() {
        return (authorities -> {
            final Set<GrantedAuthority> mappedAuthorities=new HashSet<>();
            authorities.forEach((authority)->{
                if (authority instanceof OidcUserAuthority oidcAuth){
                    mappedAuthorities.addAll(mappedAuthorities(oidcAuth.getIdToken().getClaims()));
                    System.out.println(oidcAuth.getAttributes());
                } else if (authority instanceof OAuth2UserAuthority oauth2Auth) {
                    mappedAuthorities.addAll(mappedAuthorities(oauth2Auth.getAttributes()));
                }
            });
        return mappedAuthorities;

    });
    }

    private List<SimpleGrantedAuthority> mappedAuthorities(final Map<String, Object> attributes) {
        final Map<String, Object> realmAccess= (Map<String, Object>) attributes.getOrDefault("realm_access", Collections.emptyMap());
        final Collection<String> roles=((Collection<String>)realmAccess.getOrDefault("roles", Collections.emptyList()));
        return roles.stream()
                .map(SimpleGrantedAuthority::new)
                .toList();
    }
}

`

Controller:

`@Controller 
    public class CustomerController {
        /**
         * acces a interface
         */
        private  CustomerRepository customerRepository;
        private ClientRegistrationRepository clientRegistrationRepository;
        public CustomerController(CustomerRepository customerRepository, ClientRegistrationRepository registrationRepository) {
            this.customerRepository = customerRepository;
            this.clientRegistrationRepository = clientRegistrationRepository;
        }

        @GetMapping("/customers")
        @PreAuthorize("hasAuthority('ADMIN')")//precise le role d'acces a la page
        public String customers(Model model){  //rendu html cote serveur, on a besoin du model
            List<Customer> customersList = customerRepository.findAll();
            model.addAttribute("customers", customersList);//ajout dans model d'un att customersList qui stock la liste recupere plus haut
            return "customers";//return attributeName du model=CustomerList
        }

        @GetMapping("/products")
   
        public String products(Model model){
            return "products";
        }

  
        @GetMapping("/auth")
        @ResponseBody
        public Authentication auth(Authentication authentication){
            return authentication;
        }
     
        @GetMapping("/")
        public String index(){
            return "index";
        }

        @GetMapping("/notAuthorized")
        public String notAuthorized(){
            return "notAuthorized";
        }
        @GetMapping("/oauth2Login")
        public String oauth2Login(Model model){
            String authorizationRequestBaseUri="oauth2/authorization";
            Map<String, String> oauth2AuthenticationUrls=new HashMap();
            Iterable<ClientRegistration> clientRegistrations=(Iterable<ClientRegistration>)clientRegistrationRepository;
            clientRegistrations.forEach(clientRegistration->{
                oauth2AuthenticationUrls.put(clientRegistration.getClientName(),
                        authorizationRequestBaseUri+"/"+ clientRegistration.getRegistrationId());
            });
            model.addAttribute("urls",oauth2AuthenticationUrls);
            return "oauth2Login";
        }
}
`

template:

`<!DOCTYPE html>
<html lang="en" xmlns:th="http://www.thymeleaf.org" xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout">
<head>
    <meta charset="UTF-8">
    <title>Template Customers</title>
    <link rel="stylesheet" type="text/css" href="/webjars/bootstrap/5.3.2/css/bootstrap.min.css">
    <script src="webjars/bootstrap/5.3.2/js/bootstrap.bundle.js"></script>
</head>
<body>

<nav class="navbar navbar-expand-lg bg-body-tertiary">
    <div class="container-fluid">
        <a class="navbar-brand" href="#">Navbar scroll</a>
        <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarScroll" aria-controls="navbarScroll" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="navbarScroll">
            <ul class="navbar-nav me-auto my-2 my-lg-0 navbar-nav-scroll" style="--bs-scroll-height: 100px;">
                <li class="nav-item">
                    <a class="nav-link active" aria-current="page" href="#">Home</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link" th:href="@{/customers}">Customers</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link" th:href="@{/products}">Products</a>
                </li>

                <li class="nav-item">
                    <a class="nav-link disabled" aria-disabled="true">Link</a>
                </li>
            </ul>
            <ul class="navbar">
                <li class="nav-item dropdown">
                    <a class="nav-link dropdown-toggle" href="#" role="button" data-bs-toggle="dropdown" aria-expanded="false">
                       <span th:text="${#authentication.name}">
                    </a>
                    <ul class="dropdown-menu">
                        <li><a class="dropdown-item" th:href="@{/login}">Login</a></li>
                        <li>
                     
                            <form th:action="@{/logout}" method="post">
                                <button type="submit" class="btn btn-success">Logout</button>
                            </form>
                        </li>
                        <li><hr class="dropdown-divider"></li>
                        <li><a class="dropdown-item" href="#">Profile</a></li>
                    </ul>
                </li>
            </ul>
        </div>
    </div>
</nav>

<div layout:fragment="content">

</div>
</body>
</html>`

**Comment by justin-tay** [comment]
What do you mean by logout doesn't work? What are you expecting to happen and what is happening?

There isn't anything obviously wrong with the configuration of the `OidcClientInitiatedLogoutSuccessHandler`.

When you hit your application's logout endpoint at `/logout` by default Spring will ask you to confirm the logout. If you confirm the `OidcClientInitiatedLogoutSuccessHandler` should do a `302` redirect to `http://localhost:8080/realms/sdia3/protocol/openid-connect/logout?id_token_hint=xxx&post_logout_redirect_uri=xxx`.

The browser will then hit Keycloak which then logs out the session and the redirects back to the application. You would need to ensure that the client settings in Keycloak has the `Valid post logout redirect URIs` configured properly. This is in `Clients` > `{your-client} > `Settings` under `Access Settings`.

You are attempting to delete some Keycloak cookies like `AUTH_SESSION_ID`, `KEYCLOAK_IDENTITY` and `KEYCLOAK_SESSION`. Those really aren't cookies your application can usually do anything about. In a normal deployment scenario they are unlikely to be on the same origin.

**Comment by chris-charisma** [comment]
hello, in logout I mean that when I log in I choose to log in via github, google or keycloak. If I log in, at the next login, the application is supposed to ask me to log in again to have access to the resources, in my case it is the list of clients which is in filtering mode by the controller since it is necessary Admin role to access it. However, when I log out and regardless of the login account used, it does not ask me to log in again, I automatically have access to the resources. In the demo, only JSESSIONID is cleaned at logout, I added the other cookies to test if they were the cause of the problem. Furthermore, second problem, I no longer even have access to the keycloak admin console to configure, it does not recognize any user who can connect. I can simply connect with the user created by the client on localhost8084 of the app but not on the keycloak console in localhost8080. I don't have the window that appears asking for disconnection confirmation either.

2024-04-10 14:44:12,334 WARN  [org.keycloak.events] (executor-thread-5) type="LOGIN_ERROR", realmId="e5e12b01-367e-43f6-9a86-dbd7f76d2a7a", clientId="security-admin-console", userId="null", ipAddress=
"0:0:0:0:0:0:0:1", error="user_not_found", auth_method="openid-connect", auth_type="code", redirect_uri="http://localhost:8080/admin/master/console/", code_id="d327ca45-74f0-455d-a2cd-1621083be597", username="user1"

**Reply by justin-tay** [reply]
You need to be clearer. Did you actually log out from the application at http://localhost:8084/logout? What do you mean by log in and then log in again?

**Comment by chris-charisma** [comment]
when I logout 
![image](https://github.com/keycloak/keycloak/assets/157164907/2f13ad28-f9bf-4de8-8652-504bed5b0224)
my uri becomes : http://localhost:8084/login?logout

![image](https://github.com/keycloak/keycloak/assets/157164907/c91d16db-b97e-4e6f-8c33-a0693e9b72f2)
if I clik on keycloak for login, I don't need to authenticate with my password and username, I have direct access resources server.

if I write "/logout" at my uri, I have the confirmation windows that ask me If I want logout and the logout works.
 My URI becomes yet: http://localhost:8084/login?logout **but after the windows of confirmation** to logout.

 finally the problem is not my code block:
 
![image](https://github.com/keycloak/keycloak/assets/157164907/ebfe6dc9-55e4-45b0-806c-cb0845d06b97)
![image](https://github.com/keycloak/keycloak/assets/157164907/39f55fa3-fa83-42bf-8160-b1d61041efe1)

?
if I click on my logout link, I do not have the logout confirmation window, it appears if I write myself in the /logout uri

**Reply by justin-tay** [reply]
I'm not sure why login?logout displays as it is typically disabled if a custom logout success handler is configured. Can you verify that OidcClientInitiatedLogoutSuccessHandler is being triggered and is redirecting the browser to Keycloak?

A GET on Spring Security's logout filter will trigger a confirmation page but a POST will not.

**Reply by chris-charisma** [reply]
![image](https://github.com/keycloak/keycloak/assets/157164907/4c65ace1-efb4-4dfe-9375-5a96223ab708)
Can you help me by giving me an example of how to do this, please?

**Reply by chris-charisma** [reply]
"{baseUrl}?logoutsuccess=true" this correct for you already?

**Reply by chris-charisma** [reply]
even if I do "/logout" directly, I have the logout confirmation request window, but I still have access to the resources without doing a login again, so I am not logged out

**Reply by chris-charisma** [reply]
![image](https://github.com/keycloak/keycloak/assets/157164907/4d195026-15c8-478e-895c-51072ec7d086)
my button logout work because when I click on, you can see in console that <html>

login?logout = :8084/logout

**Reply by chris-charisma** [reply]
and I click again on keycloak login and I have access to resources without a new authentication
![image](https://github.com/keycloak/keycloak/assets/157164907/9faab7b0-2e40-4701-9d20-d28cb51a31d4)

**Reply by chris-charisma** [reply]
when I'm login, I have my cookie: JSESSIONID
![image](https://github.com/keycloak/keycloak/assets/157164907/750a4b58-f280-4732-84f9-327c4c8b6961)
after logout: 
![image](https://github.com/keycloak/keycloak/assets/157164907/521d275d-c981-4469-a269-eda02800ffcd)
the cookie is not cleared by : 
 ` 
 public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http.csrf(Customizer.withDefaults())//config par def.
                .authorizeHttpRequests(ar-> ar.anyRequest().authenticated())//toutes requetes necessite une authentification
                .authorizeHttpRequests(ar->ar.requestMatchers("/","/webjars/**","/h2-console/**",  "/oauth2Login/**").permitAll())//toute url sans besoin de s'authentifier:parefeu
                .oauth2Login(al->al.loginPage("/oauth2Login"))
                .logout((logout)->logout
                        .logoutSuccessHandler(oidcLogoutSuccessHandler())
                        .logoutSuccessUrl("/").permitAll()
                        .clearAuthentication(true)
                        .deleteCookies("JSESSIONID","jti", "AUTH_SESSION_ID", "KEYCLOAK_SESSION", "KEYCLOAK_IDENTITY", "KC_RESTART"))
                        .exceptionHandling(eh->eh.accessDeniedPage("/notAuthorized"));
        return http.build();   }
 `

**Reply by justin-tay** [reply]
It's actually quite obvious that the `OidcClientInitiatedLogoutSuccessHandler` isn't being properly called. If it worked the `302` redirect should be to `http://localhost:8080/realms/sdia3/protocol/openid-connect/logout?id_token_hint=xxx&post_logout_redirect_uri=xxx` and not `http://localhost:8084/login?logout`. Also since you configured `{baseUrl}?logoutsuccess=true` as your post logout redirect that should be the final redirect uri from Keycloak to your application, but it isn't.   Unfortunately given that you have made so many customizations it's hard to determine why this is happening. You also have a `anyRequest` matcher before another more specific matcher, which shouldn't work. You should set a breakpoint in `org.springframework.security.web.authentication.logout.LogoutFilter` to see what are the handlers being called.

**Reply by chris-charisma** [reply]
ok, and I lost access to the admin console of keycloak, impossible to connect, it refuses all usernames for login like admin
